### PR TITLE
feat(vue-grid): add the Virtual Scrolling feature

### PR DESCRIPTION
**Pull request by MaximKudriavtsev**

### CODE REVIEW

File: packages/dx-vue-core/src/sizer.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
+const styles = {
```
Comment (SergeyAlexeev) [inline packages/dx-vue-core/src/sizer.js]: Can we share these styles between react and vue? Maybe use the `dx-styles` package?

File: packages/dx-vue-core/src/sizer.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    position: 'absolute',
+    top: 0,
+    left: 0,
+    width: '200%',
+    height: '200%',
+  },
+};
+
+export const DxSizer = {
+  name: 'Sizer',
```
Comment (SergeyAlexeev) [inline packages/dx-vue-core/src/sizer.js]: DxSizer

File: packages/dx-vue-grid-bootstrap4/src/plugins/virtual-table.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      type: Number,
+      default: 530,
+    },
+    headTableComponent: {
+      type: Object,
+      default: () => FixedHeader,
+    },
+    messages: {
+      type: Object,
+      default: () => { },
```
Comment (SergeyAlexeev) [inline packages/dx-vue-grid-bootstrap4/src/plugins/virtual-table.js]: `default: () => { },` return nothing. Should be `default: () => ({}),`

File: packages/dx-vue-grid-bootstrap4/src/templates/table-stub-header-cell.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export const TableStubHeaderCell = {
+  props: {
+    tableRow: {},
```
Comment (SergeyAlexeev) [inline packages/dx-vue-grid-bootstrap4/src/templates/table-stub-header-cell.js]: Use [recommended way](https://vuejs.org/v2/style-guide/#Prop-definitions-essential) for prop definitions.

File: packages/dx-vue-grid/src/components/table-layout/virtual-table-layout.test.js [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    { key: 4 },
+    { key: 5 },
+    { key: 6 },
+  ],
+  containerComponent: { render() { return <div />; } },
+  headTableComponent: { render() { return <table />; } },
+  tableComponent: { render() { return <table />; } },
+  headComponent: { render() { return <table />; } },
+  bodyComponent: { render() { return <tbody />; } },
+  rowComponent: { name: 'Row', render() { return null; } },
```
Comment (SergeyAlexeev) [inline packages/dx-vue-grid/src/components/table-layout/virtual-table-layout.test.js]: Can we use a "short" `render` function: `render: () => null`?

File: packages/dx-vue-grid/src/components/table-layout/virtual-table-layout.test.js [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+      target,
+      currentTarget: target,
+    };
+
+    // eslint-disable-next-line no-param-reassign
+    tree.find(defaultProps.containerComponent).element.target = target;
+    // eslint-disable-next-line no-param-reassign
+    tree.find(defaultProps.containerComponent).element.currentTarget = target;
+
+    tree.find(defaultProps.containerComponent).vm.$listeners.scroll(eventData);
```
Comment (SergeyAlexeev) [inline packages/dx-vue-grid/src/components/table-layout/virtual-table-layout.test.js]: vm.$emit))

File: packages/dx-vue-core/src/sizer.js (hunk 1 above)
Comment (MaximKudriavtsev) [reply packages/dx-vue-core/src/sizer.js]: I am afraid this add a new dependency for customers.
Or you mean only combine this two file to one, without add to rollup?

File: packages/dx-vue-grid/src/components/table-layout/virtual-table-layout.test.js (hunk 6 above)
Comment (MaximKudriavtsev) [reply packages/dx-vue-grid/src/components/table-layout/virtual-table-layout.test.js]: This don't work (

File: packages/dx-vue-core/src/sizer.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    return {
+      size: { width: 0, height: 0 },
+    };
+  },
+  mounted() {
+    this.setupListeners();
+  },
+  methods: {
+    setupListeners() {
+      const size = { height: this.$refs['sizer-root'].offsetHeight, width: this.$refs['sizer-root'].offsetWidth };
```
Comment (kvet) [inline packages/dx-vue-core/src/sizer.js]: Is the `sizer` prefix needed here?

File: packages/dx-vue-grid/src/components/table-layout/__snapshots__/virtual-table-layout.test.js.snap [hunk 8]
```diff
@@ -0,0 +1,7 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`VirtualTableLayout should render correct layout 1`] = `
```
Comment (kvet) [inline packages/dx-vue-grid/src/components/table-layout/__snapshots__/virtual-table-layout.test.js.snap]: Strange output

File: packages/dx-vue-grid/src/components/table-layout/virtual-table-layout.js [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+          <ColumnGroup
+            columns={collapsedGrid.columns}
+          />
+          <Body>
+            {collapsedGrid.rows.map((visibleRow) => {
+              const { row, cells = [] } = visibleRow;
+              return (
+                <DxRefHolder
+                  key={row.key}
+                  ref={`DxRefHolder-${row.key}-${row.height}`}
```
Comment (kvet) [inline packages/dx-vue-grid/src/components/table-layout/virtual-table-layout.js]: Is the `DxRefHolder` prefix really needed here?

File: packages/dx-vue-demos/src/demo-sources/grid-filtering/none/remote-filtering.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+          this.loading = false;
+        })
+        .catch(() => {
+          this.loading = false;
+        });
+      this.lastQuery = queryString;
+    },
+  },
+  template: `
+    <div class="card" style="'position': 'relative'">
```
Comment (SergeyAlexeev) [inline packages/dx-vue-demos/src/demo-sources/grid-filtering/none/remote-filtering.js]: The `.card` CSS class already contains the `position: relative` prop.

File: packages/dx-vue-demos/src/demo-sources/grid-grouping/none/remote-grouping-with-local-expanding.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+          this.tempExpandedGroups = null;
+        })
+        .catch(() => {
+          this.loading = false;
+        });
+      this.lastQuery = queryString;
+    },
+  },
+  template: `
+    <div class="card" style="'position': 'relative'">
```
Comment (SergeyAlexeev) [inline packages/dx-vue-demos/src/demo-sources/grid-grouping/none/remote-grouping-with-local-expanding.js]: The same about `.card`.

File: packages/dx-vue-demos/src/demo-sources/core-sizer/none/basic.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
+import { DxSizer } from '@devexpress/dx-vue-core';
+
+// eslint-disable-next-line react/prefer-stateless-function
```
Comment (kvet) [inline packages/dx-vue-demos/src/demo-sources/core-sizer/none/basic.js]: Remove redundant comment

File: packages/dx-vue-grid/src/components/table-layout/__snapshots__/virtual-table-layout.test.js.snap [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+<div
+  style="height: 100px;"
+>
+  <table
+    minwidth="400px"
+    style="min-width: 400px;"
+  >
+    <!---->
```
Comment (kvet) [inline packages/dx-vue-grid/src/components/table-layout/__snapshots__/virtual-table-layout.test.js.snap]: Where is the table rows and columns?

Comment (MaximKudriavtsev) [reply packages/dx-vue-grid/src/components/table-layout/__snapshots__/virtual-table-layout.test.js.snap]: Render all hierarchical is very difficult

Comment (MaximKudriavtsev) [reply packages/dx-vue-grid/src/components/table-layout/__snapshots__/virtual-table-layout.test.js.snap]: @kvet done

**Pull request merged**